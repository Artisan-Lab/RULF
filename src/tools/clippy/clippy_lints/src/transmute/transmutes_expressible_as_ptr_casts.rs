use super::utils::can_be_expressed_as_pointer_cast;
use super::TRANSMUTES_EXPRESSIBLE_AS_PTR_CASTS;
use clippy_utils::diagnostics::span_lint_and_then;
use clippy_utils::sugg;
use rustc_errors::Applicability;
use rustc_hir::Expr;
use rustc_lint::LateContext;
use rustc_middle::ty::Ty;

/// Checks for `transmutes_expressible_as_ptr_casts` lint.
/// Returns `true` if it's triggered, otherwise returns `false`.
pub(super) fn check<'tcx>(
    cx: &LateContext<'tcx>,
    e: &'tcx Expr<'_>,
    from_ty: Ty<'tcx>,
    to_ty: Ty<'tcx>,
    arg: &'tcx Expr<'_>,
) -> bool {
    if can_be_expressed_as_pointer_cast(cx, e, from_ty, to_ty) {
        span_lint_and_then(
            cx,
            TRANSMUTES_EXPRESSIBLE_AS_PTR_CASTS,
            e.span,
            &format!("transmute from `{from_ty}` to `{to_ty}` which could be expressed as a pointer cast instead"),
            |diag| {
                if let Some(arg) = sugg::Sugg::hir_opt(cx, arg) {
                    let sugg = arg.as_ty(&to_ty.to_string()).to_string();
                    diag.span_suggestion(e.span, "try", sugg, Applicability::MachineApplicable);
                }
            },
        );
        true
    } else {
        false
    }
}
