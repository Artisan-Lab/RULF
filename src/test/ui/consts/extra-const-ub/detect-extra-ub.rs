// revisions: no_flag with_flag
// [no_flag] check-pass
// [with_flag] compile-flags: -Zextra-const-ub-checks
#![feature(const_ptr_read)]

use std::mem::transmute;

const INVALID_BOOL: () = unsafe {
    let _x: bool = transmute(3u8);
    //[with_flag]~^ ERROR: evaluation of constant value failed
    //[with_flag]~| invalid value
};

const INVALID_PTR_IN_INT: () = unsafe {
    let _x: usize = transmute(&3u8);
    //[with_flag]~^ ERROR: evaluation of constant value failed
};

const INVALID_SLICE_TO_USIZE_TRANSMUTE: () = unsafe {
    let x: &[u8] = &[0; 32];
    let _x: (usize, usize) = transmute(x);
    //[with_flag]~^ ERROR: evaluation of constant value failed
};

const UNALIGNED_PTR: () = unsafe {
    let _x: &u32 = transmute(&[0u8; 4]);
    //[with_flag]~^ ERROR: evaluation of constant value failed
    //[with_flag]~| invalid value
};

const UNALIGNED_READ: () = {
    INNER; //[with_flag]~ERROR evaluation of constant value failed
    // There is an error here but its span is in the standard library so we cannot match it...
    // so we have this in a *nested* const, such that the *outer* const fails to use it.
    const INNER: () = unsafe {
        let x = &[0u8; 4];
        let ptr = x.as_ptr().cast::<u32>();
        ptr.read();
    };
};

fn main() {}
