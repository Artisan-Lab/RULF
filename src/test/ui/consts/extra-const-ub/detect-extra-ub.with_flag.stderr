error[E0080]: evaluation of constant value failed
  --> $DIR/detect-extra-ub.rs:9:20
   |
LL |     let _x: bool = transmute(3u8);
   |                    ^^^^^^^^^^^^^^ constructing invalid value: encountered 0x03, but expected a boolean

error[E0080]: evaluation of constant value failed
  --> $DIR/detect-extra-ub.rs:15:21
   |
LL |     let _x: usize = transmute(&3u8);
   |                     ^^^^^^^^^^^^^^^ unable to turn pointer into raw bytes
   |
   = help: this code performed an operation that depends on the underlying bytes representing a pointer
   = help: the absolute address of a pointer is not known at compile-time, so such operations are not supported

error[E0080]: evaluation of constant value failed
  --> $DIR/detect-extra-ub.rs:21:30
   |
LL |     let _x: (usize, usize) = transmute(x);
   |                              ^^^^^^^^^^^^ unable to turn pointer into raw bytes
   |
   = help: this code performed an operation that depends on the underlying bytes representing a pointer
   = help: the absolute address of a pointer is not known at compile-time, so such operations are not supported

error[E0080]: evaluation of constant value failed
  --> $DIR/detect-extra-ub.rs:26:20
   |
LL |     let _x: &u32 = transmute(&[0u8; 4]);
   |                    ^^^^^^^^^^^^^^^^^^^^ constructing invalid value: encountered an unaligned reference (required 4 byte alignment but found 1)

error[E0080]: evaluation of constant value failed
  --> $SRC_DIR/core/src/ptr/mod.rs:LL:COL
   |
LL |         copy_nonoverlapping(src, tmp.as_mut_ptr(), 1);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         accessing memory with alignment 1, but alignment 4 is required
   |         inside `std::ptr::read::<u32>` at $SRC_DIR/core/src/ptr/mod.rs:LL:COL
   |
  ::: $SRC_DIR/core/src/ptr/const_ptr.rs:LL:COL
   |
LL |         unsafe { read(self) }
   |                  ---------- inside `ptr::const_ptr::<impl *const u32>::read` at $SRC_DIR/core/src/ptr/const_ptr.rs:LL:COL
   |
  ::: $DIR/detect-extra-ub.rs:38:9
   |
LL |         ptr.read();
   |         ---------- inside `INNER` at $DIR/detect-extra-ub.rs:38:9

error[E0080]: evaluation of constant value failed
  --> $DIR/detect-extra-ub.rs:32:5
   |
LL |     INNER;
   |     ^^^^^ referenced constant has errors

error: aborting due to 6 previous errors

For more information about this error, try `rustc --explain E0080`.
